File: python/aurigraph_sdk/namespaces/graphql.py
"""Namespace for /api/v11/graphql -- GraphQL gateway."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from aurigraph_sdk.models import GraphQLResponse

if TYPE_CHECKING:
    from aurigraph_sdk.client import AurigraphClient


class GraphQLApi:
    """GraphQL gateway operations."""

    def __init__(self, client: AurigraphClient) -> None:
        self._client = client

    def query(
        self,
        query: str,
        variables: dict[str, Any] | None = None,
    ) -> GraphQLResponse:
        """Execute an arbitrary GraphQL query with optional variables.

        Args:
            query: The GraphQL query string.
            variables: Optional variables for the query.

        Returns:
            The GraphQL response with data and/or errors.
        """
        body: dict[str, Any] = {"query": query}
        if variables:
            body["variables"] = variables
        data = self._client._post("/graphql", body)
        return GraphQLResponse.model_validate(data)

    def query_channels(self) -> GraphQLResponse:
        """Query all channels via GraphQL."""
        return self.query(
            "{ channels { id name type status stakeholders { id role } } }"
        )

    def query_assets(self, channel_id: str) -> GraphQLResponse:
        """Query assets for a specific channel via GraphQL.

        Args:
            channel_id: The channel ID to query assets for.
        """
        return self.query(
            "query($channelId: ID!) { assets(channelId: $channelId) { id type status owner } }",
            variables={"channelId": channel_id},
        )

    def query_contracts(self) -> GraphQLResponse:
        """Query all contracts via GraphQL."""
        return self.query(
            "{ contracts { id title status templateId parties { name role } } }"
        )

    def query_node_metrics(self) -> GraphQLResponse:
        """Query node metrics via GraphQL."""
        return self.query(
            "{ nodeMetrics { totalNodes activeNodes validatorCount networkStatus } }"
        )

File: rust/src/namespaces/graphql.rs
//! GraphQL gateway namespace for `/api/v11/graphql` endpoint.
//!
//! Provides raw GraphQL queries and pre-built convenience queries
//! for common data shapes (channels, assets, contracts, node metrics).
//!
//! # Example
//!
//! ```rust,no_run
//! # async fn example(client: &aurigraph_sdk::AurigraphClient) -> aurigraph_sdk::Result<()> {
//! // Raw query
//! let result = client.graphql().query("{ channels { id name status } }", None).await?;
//! println!("Channels: {}", result);
//!
//! // Pre-built convenience query
//! let channels = client.graphql().query_channels().await?;
//! println!("Channels: {}", channels);
//! # Ok(())
//! # }
//! ```

use crate::client::AurigraphClient;
use crate::error::Result;

/// GraphQL API namespace.
///
/// Obtained via [`AurigraphClient::graphql()`].
#[derive(Debug)]
pub struct GraphQLApi<'a> {
    client: &'a AurigraphClient,
}

impl<'a> GraphQLApi<'a> {
    pub(crate) fn new(client: &'a AurigraphClient) -> Self {
        Self { client }
    }

    /// Execute a raw GraphQL query.
    ///
    /// Calls `POST /api/v11/graphql`.
    pub async fn query(
        &self,
        query: &str,
        variables: Option<&serde_json::Value>,
    ) -> Result<serde_json::Value> {
        let mut body = serde_json::json!({ "query": query });
        if let Some(vars) = variables {
            body["variables"] = vars.clone();
        }
        self.client.post("/graphql", &body).await
    }

    /// Query all channels.
    ///
    /// Pre-built convenience query: `{ channels { id name type status } }`.
    pub async fn query_channels(&self) -> Result<serde_json::Value> {
        self.query("{ channels { id name type status } }", None)
            .await
    }

    /// Query all assets with basic fields.
    ///
    /// Pre-built convenience query: `{ assets { id type status useCaseId } }`.
    pub async fn query_assets(&self) -> Result<serde_json::Value> {
        self.query("{ assets { id type status useCaseId } }", None)
            .await
    }

    /// Query all active contracts.
    ///
    /// Pre-built convenience query: `{ contracts { id template status createdAt } }`.
    pub async fn query_contracts(&self) -> Result<serde_json::Value> {
        self.query(
            "{ contracts { id template status createdAt } }",
            None,
        )
        .await
    }

    /// Query node metrics.
    ///
    /// Pre-built convenience query: `{ nodeMetrics { totalNodes activeNodes validatorCount networkStatus } }`.
    pub async fn query_node_metrics(&self) -> Result<serde_json::Value> {
        self.query(
            "{ nodeMetrics { totalNodes activeNodes validatorCount networkStatus } }",
            None,
        )
        .await
    }
}
